### PR TITLE
feat: Support for style and ansi-color in pl-code

### PR DESCRIPTION
**Pull request by jonatanschroeder**
* [x] Element
* [x] Docs
* [x] Example

### CODE REVIEW

File: elements/pl-code/pl-code.py
Comment (eliotwrobson) [inline elements/pl-code/pl-code.py:30]: If these colors are duplicated, might be a good idea to put these in a shared file somewhere?

Comment (jonatanschroeder) [reply elements/pl-code/pl-code.py:30]: Yes, I thought about that. There is a minor issue that ansi2html expects an ordered array while pygments requires a dict, so there is a need to adjust from one to the other, though this is doable.

Another option is to have pl-external-grader-result defer to pl-code for this functionality, so all colors would be found here.

Comment (eliotwrobson) [reply elements/pl-code/pl-code.py:30]: Having external grader result defer is probably the best option imo, since it should mean getting to simplify the backend of that element, similar to what we did for pl-variable-output.

Comment (jonatanschroeder) [reply elements/pl-code/pl-code.py:29]: Unfortunately `pygments-ansi-color` doesn't currently support bright colors (https://github.com/chriskuehl/pygments-ansi-color/issues/8). So we should either keep both elements separate, or lose access to bright colors in pl-external-grader-result.

Comment (jonatanschroeder) [reply elements/pl-code/pl-code.py:29]: Submitted https://github.com/chriskuehl/pygments-ansi-color/pull/27, if it is pushed and deployed relatively soon we can use it.

Comment (jonatanschroeder) [reply elements/pl-code/pl-code.py:29]: The update was merged and deployed to pip, so 8eb0187ad updates the code to use bright colors as well.

File: elements/pl-code/pl-code.py
Comment (eliotwrobson) [inline elements/pl-code/pl-code.py:27]:
```suggestion
ANSI_COLORS = {
```
Convention for class constants is to have them in all caps.

File: elements/pl-code/pl-code.py
Comment (eliotwrobson) [inline elements/pl-code/pl-code.py:162]:
```suggestion
        raise KeyError(
```

**Review by eliotwrobson — APPROVED** [review]
Minor style items but looks good overall!

File: elements/pl-code/pl-code.py
Comment (jonatanschroeder) [reply elements/pl-code/pl-code.py]: Replaced all generic Exceptions with `KeyError` or `ValueError` in 118f90313.

File: elements/pl-code/pl-code.py
Comment (jonatanschroeder) [reply elements/pl-code/pl-code.py]: Done in 118f90313.

File: elements/pl-code/pl-code.py
Comment (jonatanschroeder) [reply elements/pl-code/pl-code.py]: Moved this discussion to #7191.

File: elements/pl-code/pl-code.py
Comment (eliotwrobson) [inline elements/pl-code/pl-code.py:215]: Somewhat tangential (but we're already rearranging this file): two lines below this, I think the ending newline removal is broken, and the file doesn't have to be manually closed.

**Review by eliotwrobson — CHANGES_REQUESTED** [review]
Noticed something weird in the pl-code controller. Given the volume of changes, I think adding type hints as part of this is the way to go.

File: elements/pl-code/pl-code.py
Comment (jonatanschroeder) [reply elements/pl-code/pl-code.py:215]: Replaced with builtins in 17aa3b722.

Comment (eliotwrobson) [reply elements/pl-code/pl-code.py:215]: Are the Canadian servers on Python >= 3.9? Those builtins are from Python 3.9, but if no versioning issues I think the new code looks much cleaner?

Comment (jonatanschroeder) [reply elements/pl-code/pl-code.py:215]: 3.10.6 https://prairielearn.slack.com/archives/C0XBTPZMZ/p1673396644523079?thread_ts=1673392551.982729&cid=C0XBTPZMZ

Comment (jonatanschroeder) [reply elements/pl-code/pl-code.py:218]: It seems you're on top of it, but apparently pyright doesn't support it yet...

```
pyright 1.1.295
/home/jonatan/pl/elements/pl-code/pl-code.py
  /home/jonatan/pl/elements/pl-code/pl-code.py:221:29 - error: Cannot access member "removesuffix" for type "str"
    Member "removesuffix" is unknown (reportGeneralTypeIssues)
  /home/jonatan/pl/elements/pl-code/pl-code.py:235:39 - error: Cannot access member "removeprefix" for type "str"
    Member "removeprefix" is unknown (reportGeneralTypeIssues)
```

Comment (eliotwrobson) [reply elements/pl-code/pl-code.py:218]: Yeah, no idea why pyright doesn't pick that up, weird.

File: elements/pl-code/pl-code.py
Comment (nwalters512) [inline elements/pl-code/pl-code.py:241]: I opened https://github.com/python/typeshed/pull/9803 to remove the need for `# type: ignore` here.

File: elements/pl-code/pl-code.py
Comment (nwalters512) [inline elements/pl-code/pl-code.py:235]: Consider annotating `pl.inner_html` with a return type to remove the need for `# type: ignore` here; same with above.

**Comment by jonatanschroeder** [comment]
Probably worth adding here as well (https://github.com/PrairieLearn/PrairieLearn/issues/7191#issuecomment-1442529183):

> One issue with this is that pygments styles with dark background have typically no proper support for the default (Text) token (see https://github.com/pygments/pygments/issues/870 and https://github.com/pygments/pygments/issues/1526). I couldn't find any style with dark background and support for text (the default, non-formatted token) with a light color. So at this point I wouldn't recommend using pl-code (or anything using pygments) with a dark background unless there is an expectation that the user (instructor?) has reviewed the visibility to ensure it's good.

**Review by nwalters512 — APPROVED** [review]
Looks reasonable to me!

File: elements/pl-code/pl-code.py
Comment (eliotwrobson) [reply elements/pl-code/pl-code.py:241]: I think this is actually unavoidable, Python typechecking doesn't like dynamic class creation, see https://github.com/microsoft/pyright/issues/2058

File: elements/pl-code/pl-code.py
Comment (eliotwrobson) [reply elements/pl-code/pl-code.py:235]: That's not the issue, the `removeprefix` and `removesuffix` functions aren't getting recognized by pyright for some reason.

File: elements/pl-code/pl-code.py
Comment (nwalters512) [reply elements/pl-code/pl-code.py:241]: I tested this change locally and it worked; is there something else that you're thinking will be broken?

File: elements/pl-code/pl-code.py
Comment (jonatanschroeder) [reply elements/pl-code/pl-code.py:235]: The issue is not `pl.inner_html`, it's `removeprefix`, which is a feature introduced to `str` in Python 3.9 that apparently is not yet supported by pyright for some reason.

File: elements/pl-code/pl-code.py
Comment (eliotwrobson) [reply elements/pl-code/pl-code.py:241]: Oh interesting, I guess pyright got updated to work in these situations or something.

File: elements/pl-code/pl-code.py
Comment (eliotwrobson) [reply elements/pl-code/pl-code.py:235]: I think i know what's up actually, give me a sec to push.

Comment (nwalters512) [reply elements/pl-code/pl-code.py:235]: We pinned pyright to 3.8, so it makes sense that 3.9 features wouldn't be available: https://github.com/PrairieLearn/PrairieLearn/blob/c50f199cb622b2721bd64cb7d3b7ac55aeadb7d5/pyproject.toml#L78. Let's bump that to 3.9? That's the oldest version currently in production.

Comment (eliotwrobson) [reply elements/pl-code/pl-code.py]: @nwalters512 read my mind https://github.com/PrairieLearn/PrairieLearn/pull/7185/commits/93647d71907ba271d27938dffd9685d473b789c2

Going to go ahead and change to the simpler type annotations as well. I thought we were on 3.10 though?

Comment (nwalters512) [reply elements/pl-code/pl-code.py]: Not everywhere; I still need to find the bandwidth to get us updated in Canada.

**Comment by jonatanschroeder** [comment]
@eliotwrobson I believe I can't merge this until you make a new review, since your last review requested changes.

**Review by eliotwrobson — APPROVED** [review]

**Pull request merged**
